src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+            ProcessProjectChange(solution, projectId);
+
+            // After processing this single project, yield so the foreground thread
+            // can do more work.  Then go and loop again so we can process the 
+            // rest of the projects.
+            Task.Factory.SafeStartNew(
+                () => OnWorkspaceChangedOnForeground(cancellationToken), cancellationToken, ForegroundTaskScheduler);
+        }
+
+        private void DequeueNextProject(Solution solution, out ProjectId projectId, out bool clearAllData)
+        {
+            this.AssertIsForeground();
+
+            lock (_gate)
+            {
+                // If we detected a solution change, then add all projects to the list of projects
+                // to process.
+                clearAllData = _solutionChanged;
+                if (_solutionChanged)
+                {
+                    _changedProjects.AddRange(solution.ProjectIds);
+                }
+
+                _solutionChanged = false;
+
+                // Remove and return the first project in the list.
+                projectId = _changedProjects.FirstOrDefault();
+                _changedProjects.Remove(projectId);
+            }
+        }
+
+        private void ProcessProjectChange(Solution solution, ProjectId projectId)
+        {
+            this.AssertIsForeground();
+
+            // Remove anything we have associated with this project.
+            Dictionary<string, string> installedPackages;
+            _projectToInstalledPackageAndVersion.TryRemove(projectId, out installedPackages);
+
+            if (!solution.ContainsProject(projectId))
+            {
+                // Project was removed.  Nothing needs to be done.
+                return;
+            }
+
+            // Project was changed in some way.  Let's go find the set of installed packages for it.
+            var dteProject = _workspace.TryGetDTEProject(projectId);
+            if (dteProject == null)
+            {
+                // Don't have a DTE project for this project ID.  not something we can query nuget for.
+                return;
+            }
+
+            installedPackages = new Dictionary<string, string>();
+
+            // Calling into nuget.  Assume they may fail for any reason.
+            try
+            {
+                var installedPackageMetadata = _packageInstallerServices.GetInstalledPackages(dteProject);
+                installedPackages.AddRange(installedPackageMetadata.Select(m => new KeyValuePair<string, string>(m.Id, m.VersionString)));
+            }
+            catch
+            {
+                // TODO(cyrusn): Telemetry on this.
+            }
+
+            _projectToInstalledPackageAndVersion.AddOrUpdate(projectId, installedPackages, (_1, _2) => installedPackages);
+        }
+
+        public bool IsInstalled(Workspace workspace, ProjectId projectId, string packageName)
+        {
+            // Can be called on any thread.
```

Comment:
Or add an `ThisCanBeCalledOnAnyThread()` concept to `ForegroundThreadAffinitizedObject` [and then we could add analyzers to assert all methods in `ForegroundThreadAffinitizedObject` types declare a threading affinity kind upfront]